python/tchannel/sync/thrift.py

```diff
@@ -0,0 +1,123 @@
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+
+import inspect
+
+from tchannel.thrift.client import client_for as async_client_for
+
+
+def client_for(service, service_module, thrift_service_name=None):
+    """Build a sync client class for the given Thrift service.
+
+    The generated class accepts a TChannelSyncClient and an optional
+    hostport as initialization arguments.
+
+    Given ``CommentService`` defined in ``comment.thrift`` and registered
+    with Hyperbahn under the name "comment", here's how this might be used:
+
+    .. code-block:: python
+
+        from tchannel.sync import TChannelSyncClient
+        from tchannel.sync.thrift import client_for
+
+        from comment import CommentService
+
+        CommentServiceClient = client_for('comment', CommentService)
+
+        tchannel_sync = TChannelSyncClient('my-service')
+        comment_client = CommentServiceClient(tchannel_sync)
+
+        future = comment_client.postComment(
+            articleId,
+            CommendService.Comment("hi")
+        )
+        result = future.result()
+
+    :param service:
+        Name of the HyperBahn service being called.
+    :param service_module:
+        The Thrift-generated module for that service. This usually has
+        the same name as definied for the service in the IDL.
+    :param thrift_service_name:
+        If the Thrift service has a different name than its module, use
+        this parameter to specify it.
+    :returns:
+        An Thrift-like class, ready to be instantiated and used
+        with TChannelSyncClient.
+    """
+    assert service_module, 'service_module is required'
+    service = service or ''  # may be blank for non-hyperbahn use cases
+    if not thrift_service_name:
+        thrift_service_name = service_module.__name__.rsplit('.', 1)[-1]
+
+    method_names = [
+        name for (name, _) in inspect.getmembers(
+            service_module.Iface, predicate=inspect.ismethod
+        )
+    ]
+
+    def init(self, tchannel_sync, hostport=None, trace=False):
+        self.async_thrift = self.__async_client_class__(
+            tchannel_sync.async_client,
+            hostport,
+            trace,
+        )
+        self.threadloop = tchannel_sync.threadloop
+
+    init.__name__ = '__init__'
+    methods = {
+        '__init__': init,
+        '__async_client_class__': async_client_for(
```

Comment:
`async_client_for` generates the asynchronous class and stores it as a class-level attribute. You don't want to generate it every time you instantiate a synchronous client. This is essentially equivalent to:

    class CommentServiceSyncClient(object):
        __async_client_class__ = client_for(...)

        # ...

Again, both versions of `client_for` generate new classes at runtime. You don't need to regenerate them because their behavior is static given the previously specified parameters, so it's a good idea to save it as a class-level variable.